### PR TITLE
refactor(db): move operations into their own classes

### PR DESCRIPTION
**Pull request by kvwalker**
Fixes NODE-1897

### CODE REVIEW

File: lib/operations/execute_db_admin_command.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    super(options);
+
+    this.db = db;
+    this.selector = selector;
+  }
+
+  execute(callback) {
+    const db = this.db;
+    const selector = this.selector;
+    let options = this.options;
```
Comment (mbroadst) [inline lib/operations/execute_db_admin_command.js]: `const` ?

File: lib/operations/execute_operation_v2.js~HEAD [hunk 2]
```diff
@@ -0,0 +1,86 @@
+'use strict';
```
Comment (mbroadst) [inline lib/operations/execute_operation_v2.js~HEAD]: this file looks erroneously commited

File: test/unit/db_tests.js [hunk 3]
```diff
@@ -7,11 +7,18 @@ const sinon = require('sinon');
 class MockTopology extends EventEmitter {
   constructor() {
     super();
+    this.s = {
+      promiseLibrary: Promise
```
Comment (kvwalker) [inline test/unit/db_tests.js]: I'm not sure if this is the best way to ensure the `MockTopology` has a `promiseLibrary`, but I figured this was in the right direction.

File: lib/operations/add_user.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    let Db = loadDb();
+
+    // Did the user destroy the topology
+    if (db.serverConfig && db.serverConfig.isDestroyed()) {
+      return callback(new MongoError('topology was destroyed'));
+    }
+    // Attempt to execute auth command
+    executeAuthCreateUserCommand(db, username, password, options, (err, r) => {
+      // We need to perform the backward compatible insert operation
+      if (err && err.code === -5000) {
```
Comment (daprahamian) [inline lib/operations/add_user.js]: I'd reverse this:

```
if (!err) handleCallback(callback, err, r);
if (err.code !== -5000) return handleCallback(callback, err);

// Do the rest of the stuff here
```

File: lib/operations/add_user.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        // Fetch a user collection
+        const collection = dbToUse.collection(CONSTANTS.SYSTEM_USER_COLLECTION);
+
+        // Check if we are inserting the first user
+        count(collection, {}, finalOptions, (err, count) => {
+          // We got an error (f.ex not authorized)
+          if (err != null) return handleCallback(callback, err, null);
+          // Check if the user exists and update i
+          const findOptions = Object.assign({ projection: { dbName: 1 } }, finalOptions);
+          collection.find({ user: username }, findOptions).toArray(err => {
```
Comment (daprahamian) [inline lib/operations/add_user.js]: Just realized: does this require passing a session in?

File: lib/operations/connect.js [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+
+class ConnectOperation extends OperationBase {
+  constructor(mongoClient, err) {
+    super();
+
+    this.mongoClient = mongoClient;
+    this.err = err;
+  }
+
+  static get aspects() {
```
Comment (daprahamian) [inline lib/operations/connect.js]: should probably be replaced with a `defineAspects` call

File: lib/operations/drop_collection.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+const OperationBase = require('./operation').OperationBase;
+const CommandOperation = require('./command');
+const handleCallback = require('../utils').handleCallback;
+const MongoError = require('mongodb-core').MongoError;
+
+class DropCollectionOperation extends OperationBase {
```
Comment (daprahamian) [inline lib/operations/drop_collection.js]: extends CommandOperation?

File: lib/operations/drop_database.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    this.command = command;
+  }
+
+  execute(callback) {
+    const db = this.db;
+    const command = this.command;
+    const options = this.options;
+
+    const commandOperation = new CommandOperation(db, command, options);
+    commandOperation.execute((err, result) => {
```
Comment (daprahamian) [inline lib/operations/drop_database.js]: same here. extend CommandOperation

File: lib/operations/profiling_level.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    super(options);
+
+    this.db = db;
+  }
+
+  execute(callback) {
+    const db = this.db;
+    const options = this.options;
+
+    const commandOperation = new CommandOperation(db, { profile: -1 }, options);
```
Comment (daprahamian) [inline lib/operations/profiling_level.js]: extend CommandOperation

File: lib/operations/add_user.js (hunk 5 above)
Comment (kvwalker) [reply lib/operations/add_user.js]: I'm not sure, does it?

**Pull request merged**